src/NUnitFramework/framework/Constraints/UniqueItemsConstraint.cs

```diff
@@ -52,97 +55,112 @@ public override string Description
         /// <returns></returns>
         protected override bool Matches(IEnumerable actual)
         {
-            return TryFastAlgorithm(actual) ?? OriginalAlgorithm(actual);
+            var nonUniqueItems = TryFastAlgorithm(actual) ?? OriginalAlgorithm(actual);
+            return nonUniqueItems.Count == 0;
+        }
+
+        /// <inheritdoc />
+        public override ConstraintResult ApplyTo<TActual>(TActual actual)
+        {
+            IEnumerable enumerable = ConstraintUtils.RequireActual<IEnumerable>(actual, nameof(actual));
+
+            var nonUniqueItems = TryFastAlgorithm(enumerable) ?? OriginalAlgorithm(enumerable);
+
+            return new UniqueItemsContstraintResult(this, actual, nonUniqueItems);
         }
 
-        private bool OriginalAlgorithm(IEnumerable actual)
+        private ICollection OriginalAlgorithm(IEnumerable actual)
         {
-            var list = new List<object>();
+            var nonUniques = new List<object>();
+            var processedItems = new List<object>();
 
-            foreach (object o1 in actual)
+            foreach (var o1 in actual)
             {
-                foreach (object o2 in list)
+                var isUnique = true;
+                var unknownNonUnique = false;
+
+                foreach (var o2 in processedItems)
+                {
                     if (ItemsEqual(o1, o2))
-                        return false;
-                list.Add(o1);
+                    {
+                        isUnique = false;
+                        unknownNonUnique = !nonUniques.Any(o2 => ItemsEqual(o1, o2));
```

Comment:
I suspect this will be contributing to the performance test needing tweaking. Here, you're checking the entire list of every non unique item for a match, for every new item you find. Why not just add them all to a list, and then `.Distinct()` at the end? 🙂

Reply:
Or - taking inspiration from the version before, add them to a Hashmap. On false, add the non-unique items to a second hashmap, and check the size of that second hash map. End the method whenever you've run out of items to check, or found the first 10 non-unique items - whichever comes first?

Reply:
Thanks @ChrisMaddock ! Things are a bit busy for me right now, but I'll slowly try and poke away at this over the next little bit.

Reply:
Turns out I had a bit more time than expected. 😄 

I tried using `HashSet<T>` or `Distinct<T>()` here, but ran into issues because this path deals with NUnitEqualityComparer (including the scenario of having user-defined comparers or equality between different types, like `42` and `42f`). Trying to wrap `ItemsEqual` into a `IEqualityComparer<object>` didn't seem to pan out perfectly, since objects with equality (like `42` and `42f`) had different default hash codes and often ended up in different hash buckets. This meant "equal" objects weren't flagged, and uniques ended up getting missed.

I think I did make some progress by adding a fastpath for when all the items implement `IComparable`, which let me then use a `SortedSet<T>`, resulting in O(n log n) instead of the original O(n^2). The performance test cases now pass 😁

Reply:
If the items implement IComparable, how do you know that the implementation is consistent with the custom equality comparer that was set in the assertion?

Reply:
Thanks @jnm2 !

For the IComparable path, `ItemEquals` is used to check if the items are equal first (which will use NUnitEqualityComparer, including any "ExternalComparers") and if they are unequal, _then_ try to apply some sort of sort based on the IComparable. I think I recently added a string-specific case to handle if `IgnoreCase` is specified, though I should update that to fall back to string checking _after_ the call to the NUnitEqualityComparer:

https://github.com/nunit/nunit/pull/3523/files#diff-f6b841c6c7eed9838e33cf9082f26944R268-R271

Sorry, I'm unsure if this answers your question

Reply:
On the note of checking for distincts while processing or at the end, I did a quick benchmark using int arrays:

<pre>
|                  Method |     N |         Mean |       Error |      StdDev | Ratio | RatioSD |
|------------------------ |------ |-------------:|------------:|------------:|------:|--------:|
|                  UseSet |    10 |     314.2 ns |     2.22 ns |     2.08 ns |  1.00 |    0.00 |
| UseHashSet_PostDistinct |    10 |     373.6 ns |     6.36 ns |    12.10 ns |  1.18 |    0.05 |
|                         |       |              |             |             |       |         |
|                  UseSet |   100 |   2,562.1 ns |    12.98 ns |    10.84 ns |  1.00 |    0.00 |
| UseHashSet_PostDistinct |   100 |   2,886.5 ns |    56.10 ns |    62.36 ns |  1.13 |    0.02 |
|                         |       |              |             |             |       |         |
|                  UseSet | 10000 | 309,340.6 ns | 2,704.61 ns | 2,397.57 ns |  1.00 |    0.00 |
| UseHashSet_PostDistinct | 10000 | 326,946.5 ns | 6,145.89 ns | 6,036.08 ns |  1.06 |    0.02 |
</pre>

Reply:
Ah, I follow now. Thanks! For what types does IComparable/SortedSet lend a solution that IEquatable/HashSet doesn't? `42` vs `42f` should be a problem for both, right?

Reply:
It has less to do with the types themselves and most to do with the containing data structure. Based on the pre-PR state of the constraint, a HashSet-based "fast-path" was used if:
- No "ExternalComparer" was specified
- A generic container (ex: List<T>) was passed
- typeof(T) is sealed
- T is not a type with special handling by NUnit (list here: https://github.com/nunit/nunit/blob/master/src/NUnitFramework/framework/Constraints/UniqueItemsConstraint.cs#L153-L158)

So if a collection of type `List<object>` or `IList` were passed, it would have to go down the "slow path", regardless of containing data (even if they were all the same data type, like `int`). The IComparable/SortedSet was my attempt to make this "slow path" a little more intelligent by checking if the data is sortable (from which it becomes much simpler to identify duplicate data, even if the container has multiple different data types).

I'm still not sure if this is the absolute best way to have approached it, so I'd welcome any feedback or suggestions on it. :)

Reply:
As part of my latest push, I've added tests for this "medium path", which uses a container which only implements the non-generic `ICollection`

Reply:
What are the real-world cases in particular that this optimization is addressing? I wouldn't have expected List<object> containing all elements of the same type to come up often. Is it mainly about `typeof(T)` being sealed?

Reply:
Sorry, yes, I oversimplified with my example of `List<object>`.

One (pre-existing) condition which will cause the fastpath to not be taken is if `typeof(T)` is unsealed. So if one passes a `List<T>` where T is unsealed but `typeof(T)` implements IComparable (or all contained items derived from T happen to implement IComparable), then this new "medium path" would be taken instead. Similarly, if a non-generic collection like `IList` or `ICollection` were passed, but all items implement IComparable, it would take the medium path.

That last case was the important one for me, as my day job involves maintaining a codebase originating from the .NET20 days, with many collections of type `List<object>` (originally these were  `IList`, but they were updated to `List<object>` during a past upgrade since both enumerate `object`s).

I know it's a bit ugly to have used the NET35 conditional compilation. If you think this optimization is safe and makes sense to keep, I'm happy to try a few different other implementations.

Reply:
My outstanding concern is that the implementation of IComparable.CompareTo may contradict NUnit's equality comparer.

What do you think about the approach of scanning every item in the collection to determine if NUnit equality *would* delegate to the type's .Equals method and then using a hash set if so? Would that address your scenario?

Reply:
I'm not entirely sure I understand, but do you mean apply the check here to every item's type in the `IEnumerable`?
https://github.com/nunit/nunit/blob/15175cba113b3937862d2f258320429658ce7440/src/NUnitFramework/framework/Constraints/UniqueItemsConstraint.cs#L82

If so, that's a really good idea. I can give that a try.

Reply:
Thanks for the suggestion! This worked. I've done this in the [TryInferFastPath](https://github.com/nunit/nunit/pull/3523/files#diff-f6b841c6c7eed9838e33cf9082f26944R104) function below.